### ISSUE TITLE
[bitnami/redis-cluster] Adding a new node to the cluster only add master node

### ISSUE THREAD

**Issue by qi-min** [bitnami/charts#3743]
**Which chart:**
[bitnami/redis-cluster]

**Describe the bug**
when I run helm upgrade to add node, the new added nodes are all master nodes, no replicas nodes added

**To Reproduce**

Helm install redis-cluster with default settings
Helm upgrade with the following settings:
```
helm upgrade --timeout 600s redis  --set "password=${REDIS_PASSWORD},cluster.nodes=8,cluster.update.addNodes=true,cluster.update.currentNumberOfNodes=6,cluster.init=false" bitnami/redis-cluster
```
Expected behavior
two new nodes added, one is master and one is slave. There should be 4 master and 4 slave node after upgrade.

**Comment by javsalgar** [comment]
Hi,

Thank you very much! I was able to reproduce the issue. I will open a task for investigation. As soon as we have more news, I will update this ticket.

**Comment by rxsto** [comment]
We're trying to use a custom scheduler to bind the redis nodes to high available zones and are experiencing the same issue. All the nodes (including the new ones) turn into standalone masters and the job keeps crashing since all the available slots are not covered. We're trying to deploy a high available redis cluster with improved failover handling, further investigation would be appreciated! Thanks :)

**Comment by javsalgar** [comment]
Thanks for the input! I will forward this to the rest of the team

**Comment by rxsto** [comment]
So, we've tried to validate the fix that was merged with #3928 but it seemingly still didn't work.

Here's what we've tried:

- Launch a testing kubernetes cluster with 3 nodes and 3 masters.
- Installed a redis cluster with 3 masters and 3 slaves (worked fine, the cluster state was healthy)
- Tested failover, restarted one kubernetes node, redis cluster successfully handled it and elected a new master for the failing pair
- Rebooted the node, manually re-elected the old master, cluster still healthy and working
- Added a new kubernetes node to the cluster
- Changed the values to include 2 new redis nodes in the next helm upgrade
- Upgraded the helm deployment, 2 new nodes were added to the cluster
- The whole statefulset restarted one pod after another just how it's used to act
- After the rollout was done, every single redis node was a single master in a different redis cluster, there was absolutely no communication between the nodes, they seemingly forgot their old role

We're not using a LoadBalancer for the redis cluster service, and we're also not using persistent volumes, since the redis cluster should definitely be able to keep its keys and values while having a minor outage or updating the redis cluster.

We're also using a custom scheduler to make sure the redis nodes are evenly distributed and automatic failover handling actually works. (`(podId % k8sNodeCount) + smallestK8sNodeId`)

We really hoped the pull request mentioned above would fix this issue, but we're still not able to maintain a healthy redis cluster with this chart.

If we're missing anything, please let us know! That would be highly appreciated!

Thanks

**Comment by rafariossaa** [comment]
Hi,
  Thanks for your thorough testing. It is only failing for you when adding nodes to the cluster ? I mean, Does in the rest of the cases the cluster was with all the nodes ?

**Comment by rxsto** [comment]
Hey,

I've just been trying to reproduce this issue without adding new kubernetes nodes, and from what I can tell the redis cluster always gets destroyed and ends up in a failing state when upgrading the helm release.

I tried to use the command `helm upgrade --timeout 600s redis-cluster --set "password=password,cluster.nodes=8,cluster.update.addNodes=true,cluster.update.currentNumberOfNodes=6" bitnami/redis-cluster -n redis-cluster` but ended up getting `Error: UPGRADE FAILED: cannot patch "redis-cluster" with kind StatefulSet: StatefulSet.apps "redis-cluster" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden`. Changing the values.yaml file and upgrading the helm release with the new files forces the cluster to fall apart. New added nodes seem to work at first, but once the rollout finished all of the old nodes are masters in their own redis cluster, just as I described in my earlier comment.

So to sum this up, updating/upgrading the helm release always destroys the redis cluster, with and without adding new kubernetes nodes.

**Comment by rafariossaa** [comment]
Hi, 
  Sorry for the long delay.
  I am adding all this notes to a internal task to dig deeper into this issue.

**Comment by fantariaYY** [comment]
Hi,

We are having a similar issue when deploying a helm upgrade (adding more Redis node, for instance) exactly as @rxsto described above, and I did some investigations on it.
Basically after helm upgrade deployment, the entire statefulset gets a rolling restart. During this period, I noticed that the restarted Redis Pod will not retain its original 40-digit uuid, therefore not getting added back to the cluster. 
Here is a walkthrough of how I did the testing:
1. Start a Redis cluster with 9 node (3 masters + 6 slaves). The cluster started up fine without any issue.
2. Deploy a helm upgrade to add the cluster to 12 nodes with the following command:

> helm3 upgrade redis-cluster objects/namespaces/redis/redis-cluster/  --namespace redis --set "cluster.nodes=12,cluster.update.addNodes=true,cluster.update.currentNumberOfNodes=9,cluster.init=false"
The helm grade successfully create the instances and the clusters was properly updated to 4 master and 8 slaves.
3. The statefulset starts to rolling restart the existing 9 instances one by one. Every restarted instances loses its original UUID and never gets added back to the cluster. The newly created 3 nodes will be a cluster of its own with 9 nodes failing.

This is an restarted pod: (pod/redis-cluster-6)
> 127.0.0.1:6379> cluster nodes
831839d95e5a0219a34069fb64f64aaa72c0d46b 10.29.250.204:6379@16379 myself,master - 0 0 0 connected 

This is a newly added pod: (pod/redis-cluster-11)

> 127.0.0.1:6379> cluster nodes
d794d4ddaf3462fa3a0d98b29132a8887f42742e 10.29.249.147:6379@16379 slave,fail e1dd4a291a99ae4e2ab8e402100d65198bd1f547 1607374514647 1607374512000 1 connected
b44e75c1f1ec43de13bed9c5f4e7004aad382328 :0@0 slave,fail,noaddr 47295f774dbd6ba33d22af8ce2b11712a5da9343 1607374537591 1607374530000 3 disconnected
e1dd4a291a99ae4e2ab8e402100d65198bd1f547 10.29.251.248:6379@16379 master,fail? - 1607374640892 1607374636887 1 connected 0-5460
12876b243a24cc3c57e2a9d7dc023f1f106172c6 :0@0 slave,fail,noaddr e1dd4a291a99ae4e2ab8e402100d65198bd1f547 1607374494608 1607374491000 1 disconnected
5cc357e4679e4e6222c502f963ecfe2558870088 10.29.251.63:6379@16379 master,fail - 1607374614842 1607374611000 2 connected 5461-10922
47295f774dbd6ba33d22af8ce2b11712a5da9343 :0@0 master,fail,noaddr - 1607374573759 1607374570000 3 disconnected 10923-16383
34e5760aba29a8c47c511b9a77c4db46f37ee8c3 10.29.248.153:6379@16379 myself,slave 5cc357e4679e4e6222c502f963ecfe2558870088 0 1607376726000 2 connected
13731b46fd4350ef47329db960d5c47bc69fda9c :0@0 slave,fail,noaddr - 1607374418465 0 0 disconnected
b94960f7426e95373901af9b6b614b776951b685 10.29.249.241:6379@16379 slave 72721c47f11de834e58deedf3a88495e1783cd43 0 1607376734059 10 connected
72721c47f11de834e58deedf3a88495e1783cd43 10.29.251.202:6379@16379 master - 0 1607376733000 10 connected
da569e7b3f9c644c17fcf7b39b425f478c8a6894 :0@0 slave,fail,noaddr 5cc357e4679e4e6222c502f963ecfe2558870088 1607374439504 1607374432000 2 disconnected
777c51a1211fa8f3af62dc7b3b8921637d2ef6fb :0@0 slave,fail,noaddr 5cc357e4679e4e6222c502f963ecfe2558870088 1607374475575 1607374470000 2 disconnected

**Comment by srfaytkn** [comment]
Any progress?

**Comment by rafariossaa** [comment]
Hi,
  I am sorry, this task stills in our backlog.

**Comment by ItsReddi** [comment]
It seems that not only that is broken on adding new nodes.
In our case it starts the cluster update, adding new nodes.
Cluster Update finish with success:
```
[OK] All nodes agree about slots configuration.

>>> Check for open slots...

>>> Check slots coverage...

[OK] All 16384 slots covered.

>>> Send CLUSTER MEET to node 10.42.56.7:6379 to make it join the cluster.

Waiting for the cluster to join

>>> Configure node as replica of 10.42.56.6:6379.

[OK] New node added correctly.
```
after that a new cluster update spawns: (it will never finish)
```
[OK] All nodes agree about slots configuration.

>>> Check for open slots...

>>> Check slots coverage...

[OK] All 16384 slots covered.

[ERR] Node 10.42.56.6:6379 is not empty. Either the node already knows other nodes (check with CLUSTER NODES) or contains some key in database 0.

Add-node 12 10.42.56.6 failed, retrying
```

and the new nodes are hanging around without any shards:
```
a2f32b4f1f7fbe5c5ca06710b4e3ad0b99f3748c 10.42.32.2:6379@16379 slave 4192b19ac21fa29b677d63e5425b7d41eb7859e1 0 1610626139590 13 connected
ee2627270fe9d8aabd9b90936d7bcadab332a802 10.42.56.7:6379@16379 slave 241f5fdccfad9794b4aa14de0f908f1217be609c 0 1610626140893 0 connected
7e7ea433972db4ef6179253e267a22154c3826bf 10.42.32.5:6379@16379 master - 0 1610626140000 15 connected 5461-8191
112cede46b55767881ad3666f5dc6e6454609469 10.42.128.5:6379@16379 master - 0 1610626138000 6 connected 13653-16383
6b2e7cb5c7c532e4eface772914f58da4140892c 10.42.16.2:6379@16379 slave f3af639a4c610e7f76a7fddbaf8384a2c9bce4ac 0 1610626135000 2 connected
fee54d26f63d154f4dcacda2acafd3c58042fdde 10.42.160.8:6379@16379 slave 112cede46b55767881ad3666f5dc6e6454609469 0 1610626134000 6 connected
15e82c165b26a0e506965452819a83d7daed2127 10.42.128.10:6379@16379 master - 0 1610626136000 14 connected 8192-10922
4192b19ac21fa29b677d63e5425b7d41eb7859e1 10.42.64.9:6379@16379 master - 0 1610626139890 13 connected 10923-13652
7cc897797d7a6974201d32231e7e191bd8ec56ae 10.42.64.2:6379@16379 master - 0 1610626138000 16 connected 0-2730
241f5fdccfad9794b4aa14de0f908f1217be609c 10.42.56.6:6379@16379 myself,master - 0 1610626136000 0 connected
f3af639a4c610e7f76a7fddbaf8384a2c9bce4ac 10.42.128.3:6379@16379 master - 0 1610626137000 2 connected 2731-5460
f9a0aec8dba72eae7c912b99640829de6eb97a4e 10.42.160.6:6379@16379 slave 7e7ea433972db4ef6179253e267a22154c3826bf 0 1610626138889 15 connected
a427e307b1972098cfbae7820ead0fb2bb3795a4 10.42.64.7:6379@16379 slave 15e82c165b26a0e506965452819a83d7daed2127 0 1610626139000 14 connected
075209408c7829c74d7ec23fe67ef93cdb9c2644 10.42.32.1:6379@16379 slave 7cc897797d7a6974201d32231e7e191bd8ec56ae 0 1610626137877 16 connected
```

**Comment by rafariossaa** [comment]
Hi,
  Thanks for providing more feedback. I am adding this information to our internal task.

**Comment by OpsPita** [comment]
Is there any new about this issue?

**Comment by rafariossaa** [comment]
Hi,
  I tried again with the latest chart version 5.0.0, and I upgraded from 6 to 8 nodes without any issues.
  I cloned the repository and used the default falues.

```console
$ helm install myredisclus -f values.yaml .

$ kubectl exec -it pod/myredisclus-redis-cluster-5 bash
I have no name!@myredisclus-redis-cluster-0:/$ redis-cli -c -h localhost  -a $REDIS_PASSWORD                            
Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
localhost:6379> 

localhost:6379> CLUSTER INFO
cluster_state:ok
cluster_slots_assigned:16384
cluster_slots_ok:16384
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:6
cluster_size:3
cluster_current_epoch:6
cluster_my_epoch:2
cluster_stats_messages_ping_sent:22
cluster_stats_messages_pong_sent:19
cluster_stats_messages_meet_sent:1
cluster_stats_messages_sent:42
cluster_stats_messages_ping_received:19
cluster_stats_messages_pong_received:23
cluster_stats_messages_received:42
l

ocalhost:6379> CLUSTER NODES
1e4da28cebceb58c9088decacb100ec74f0fdb13 10.168.0.10:6379@16379 master - 0 1618824505000 3 connected 10923-16383
4e743eeb932746b3df19ad34d61e11bd2d64a9fa 10.168.2.13:6379@16379 slave 1e4da28cebceb58c9088decacb100ec74f0fdb13 0 1618824504000 3 connected
3b8b1137958a9a4ca6090eef9e6673b83314278e 10.168.2.14:6379@16379 myself,slave e9f8934ad125e54fc703f0cf4be6cbcaf8ad4593 0 1618824503000 2 connected
cc75bfa9a9d31abcfd45e4d72a39396dd850ddac 10.168.0.9:6379@16379 slave 54e9fe09cc652ade93548b00699f8c8af3b6fe48 0 1618824506993 1 connected
e9f8934ad125e54fc703f0cf4be6cbcaf8ad4593 10.168.1.11:6379@16379 master - 0 1618824505000 2 connected 5461-10922
54e9fe09cc652ade93548b00699f8c8af3b6fe48 10.168.1.10:6379@16379 master - 0 1618824506000 1 connected 0-5460
localhost:6379> 
```

  Then I upgraded to 8 nodes:

```console
$  helm upgrade myredisclus  --set "password=${REDIS_PASSWORD},cluster.nodes=8,cluster.update.addNodes=true,cluster.update.currentNumberOfNodes=6,cluster.init=false" -f values.yaml .
```

  I will add the new pods and recreate the previous ones. After it, I got all the nodes in master/slaves pairs.

```console
$ kubectl exec -it pod/myredisclus-redis-cluster-5 bash
I have no name!@myredisclus-redis-cluster-5:/$ redis-cli -c -h localhost  -a $REDIS_PASSWORD
Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
localhost:6379> 

localhost:6379> CLUSTER INFO
cluster_state:ok
cluster_slots_assigned:16384
cluster_slots_ok:16384
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:8
cluster_size:3
cluster_current_epoch:7
cluster_my_epoch:7
cluster_stats_messages_ping_sent:289
cluster_stats_messages_pong_sent:276
cluster_stats_messages_auth-req_sent:5
cluster_stats_messages_sent:570
cluster_stats_messages_ping_received:271
cluster_stats_messages_pong_received:284
cluster_stats_messages_fail_received:3
cluster_stats_messages_auth-ack_received:2
cluster_stats_messages_received:560

localhost:6379> CLUSTER NODES
e9f8934ad125e54fc703f0cf4be6cbcaf8ad4593 10.168.1.13:6379@16379 slave 3b8b1137958a9a4ca6090eef9e6673b83314278e 0 1618824831000 7 connected
cc75bfa9a9d31abcfd45e4d72a39396dd850ddac 10.168.0.12:6379@16379 slave 54e9fe09cc652ade93548b00699f8c8af3b6fe48 0 1618824831000 1 connected
1e4da28cebceb58c9088decacb100ec74f0fdb13 10.168.0.13:6379@16379 master - 0 1618824832085 3 connected 10923-16383
54e9fe09cc652ade93548b00699f8c8af3b6fe48 10.168.1.14:6379@16379 master - 0 1618824830081 1 connected 0-5460
3b8b1137958a9a4ca6090eef9e6673b83314278e 10.168.2.16:6379@16379 myself,master - 0 1618824829000 7 connected 5461-10922
dbf6b215fd54ec58cad4346d70b378de1a4bc101 10.168.0.11:6379@16379 slave 90d132a315db72b5b69841a5469faeed2d84a94b 0 1618824832000 0 connected
90d132a315db72b5b69841a5469faeed2d84a94b 10.168.1.12:6379@16379 master - 0 1618824832000 0 connected
4e743eeb932746b3df19ad34d61e11bd2d64a9fa 10.168.2.17:6379@16379 slave 1e4da28cebceb58c9088decacb100ec74f0fdb13 0 1618824833086 3 connected
```

  Could you try with the latest version ?
